allennlp/nlp_api.py

```diff
@@ -0,0 +1,289 @@
+        ``recurrent_layers``, the names correspond to places in your model where these cells get
+        used, not to anything about what kind of cell you're using.
+    default_recurrent_cell : ``Callable[[], torch.nn.Module]``, optional (default=``None``)
+        A function that returns a new recurrent cell.  Giving this makes it so you can use unique
+        names for all recurrent cells in your model, but still only have to specify one cell type.
+        See :func:`get_recurrent_cell` for more detail.
+    """
+    def __init__(self,
+                 embedders: Dict[str, Union[torch.nn.Module, List[torch.nn.Module]]] = None,
+                 recurrent_layers: Dict[str, torch.nn.Module] = None,
+                 default_recurrent_layer: Callable[[], torch.nn.Module] = None,
+                 recurrent_cells: Dict[str, torch.nn.Module] = None,
+                 default_recurrent_cell: Callable[[], torch.nn.Module] = None):
+        if embedders is None:
+            embedders = {}
+        self._embedders = embedders
+        if recurrent_layers is None:
+            recurrent_layers = {}
+        self._recurrent_layers = recurrent_layers
+        self._default_recurrent_layer = default_recurrent_layer
+        if recurrent_cells is None:
+            recurrent_cells = {}
+        self._recurrent_cells = recurrent_cells
+        self._default_recurrent_cell = default_recurrent_cell
+
+    def embed_text_field(self,
+                         input_tensors: List[torch.Tensor],
+                         name: str = "default",
+                         fallback_behavior: str = "crash") -> torch.Tensor:
+        """
+        Given an input corresponding to a :class:`~allennlp.data.fields.TextField`, return an
+        embedding of the tokens in that field.  The ``name`` here is typically the name of an input
+        field, like "passage", or "question".  And we will use the ``embedders`` specified for the
+        given ``name`` in the constructor to actually perform the embedding.
+
+        ``TextFields`` are represented in AllenNLP as `lists` of tensors, where each item in the
+        list is one representation of the tokens in the ``TextField`` (e.g., word IDs, or character
+        IDs per word).  This lets us be flexible in specifying the model, changing the
+        representation used for ``TextFields`` in your data, and the way that those representations
+        are embedded, without changing any model code at all.  The recommended way to leverage this
+        abstraction in a model like Bidirectional Attention Flow, which does passage-based question
+        answering, is something like this::
+
+            # In the model code.  This assumes a data type that has four ``Fields``: "passage",
+            # "question", "span_begin", and "span_end".
+            def forward(self, inputs):
+                passage_tokens = inputs['passage']
+                embedded_passage = nlp_api.embed_text_field(passage_tokens)
```

Comment:
It's not, actually - if you always want to embed `TextFields` the same way in your model, you can just leave this as it is, and specify an embedder with the `default` key in the constructor.  I'll make this more clear with some better documentation for each method, as you recommended for the class docstring.